### PR TITLE
Plugins: Show upload page at /plugins/upload

### PR DESCRIPTION
**Pull request by seear**
Add a page at /plugins/upload/{site} that allows plugin zips to be uploaded to a jetpack site.

Addresses #15664. Full flow design: p6TEKc-1j6-p2

screen shot 2017-07-19 at 16 02 32
screen shot 2017-07-19 at 16 02 53
screen shot 2017-07-19 at 16 03 05

**To Test**
* Go to http://calypso.localhost:3000/plugins/upload/{jetpacksite} (or use calypso.live)
* Drag a plugin zip over the drop component, or click to pick with a file browser

**Expected:**
* Progress bar should move for file upload then pulse whilst waiting for response from site
* Any error (for example, plugin already installed on site) will not show. Notices will be added in a further PR
* Successful upload will result in the plugin page loading
* Custom plugins currently only display name and author

### CODE REVIEW

File: client/my-sites/plugins/plugin-upload/index.jsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	back = () => {
+		page.back();
+	}
+
+	render() {
+		const { translate } = this.props;
+
+		return (
+			<Main>
+				<HeaderCake onClick={ this.back }>{ translate( 'Upload plugin' ) }</HeaderCake>
```
Comment (a8ci18n) [inline client/my-sites/plugins/plugin-upload/index.jsx]: Hi! I've found a possible matching string that has already been translated **68** times:
`translate( 'Upload Plugin' )` *ES Score: 15*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/plugins/plugin-upload/index.jsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+				</FilePicker>
+			</div>
+		);
+	}
+
+	renderProgressBar() {
+		const { translate, progress, installing } = this.props;
+
+		const uploadingMessage = translate( 'Uploading your plugin' );
+		const installingMessage = translate( 'Installing your plugin' );
```
Comment (a8ci18n) [inline client/my-sites/plugins/plugin-upload/index.jsx]: Hi! I've found a possible matching string that has already been translated **53** times:
`translate( 'Installing the plugin&#8230;' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/plugins/plugin-upload/index.jsx (hunk 1 above)
Comment (seear) [inline client/my-sites/plugins/plugin-upload/index.jsx]: 👍

**Comment by drw158** [comment]
So right now, you have to type in the URL to access this page? Any plans for adding a way to access it in the UI?

**Comment by folletto** [comment]
Design wise looks good.

I think I've a question too. :D 
I was expecting an action forward on the final page... any plan for that too? :)

**Comment by seear** [comment]
Sorry, poor background from me in the PR description. Updated to include a link to the latest design: p6TEKc-1j6-p2

> So right now, you have to type in the URL to access this page? Any plans for adding a way to access it in the UI?

Yes, an _Upload_ button in various places, see design above

> I was expecting an action forward on the final page... any plan for that too? :)

Yes, an _Activate_ button will be prominent. This just uses the standard plugin view for now.

**Comment by folletto** [comment]
Ok, cool then! 👍

**Comment by keoshi** [comment]
@seear Did some more testing with a couple of different setups and it's not working for multi-site or Atomic sites — do we support those?

**Comment by seear** [comment]
> not working for multi-site or Atomic sites — do we support those?

With multi-site I believe there are some technical limitations. Theme upload for multisite is disabled, and I suggest we do the same to start with for plugins. Tracking in #15664 (I'll look into the technical detail further and expand on it).

For (already) Atomic sites, this should be working. It worked ok for one of my sites just now. When we add error notices soon we can see what the problem is with your site.

screen shot 2017-07-20 at 11 07 40

**Comment by keoshi** [comment]
Re: multisite: true, we should launch those in parallel when it happens.

When testing in my Atomic site, it didn't work. I saw the progress bar (screen no. 2), but was redirected to the plugin upload (screen no. 1). Curious to know the exact error, since it worked with theme uploads just a few weeks ago.

**Comment by sirreal** [comment]
> Curious to know the exact error, since it worked with theme uploads just a few weeks ago.

@keoshi If you open the network tab, you should be able to see the error even though it's hidden in this PR:

error-msg

**Comment by keoshi** [comment]
@sirreal Cool, thank you!

Got the same error:

image

**Comment by seear** [comment]
`folder_exists` means that plugin is already installed on the site.

**Pull request merged**